### ISSUE TITLE
_get_any expecting Last-Modified

### ISSUE THREAD

**Issue by incubus005** [uktrade/jupyters3#4]
_get_file_base64(context, path='',content=False) calls _get_any
_get_any is expecting Last-Modified key in response headers:
```python
last_modified_str = response.headers['Last-Modified']
 ```

but the response example headers from  HEAD https://{aws_s3_host}/{aws_s3_bucket}/:
X-Amz-Id-2: wGFCYsbbV6HdvYBW10exxXVJ3YQKoD38Tt//qa84nDy+FZ0DG0Pwq7PVMNTA1Wvrwn8Ad4Xhp
X-Amz-Request-Id: 7DF61C1CBB3FE8
Date: Tue, 23 Apr 2019 21:20:24 GMT
X-Amz-Bucket-Region: eu-west-1
Content-Type: application/xml
Transfer-Encoding: chunked
Server: AmazonS3
Connection: close

```[E 22:20:23.285 NotebookApp] Uncaught exception GET /tree?token=c0949919ea655c22346df6765b47e8ec323eff55febe3128 (::1)
    HTTPServerRequest(protocol='http', host='localhost', method='GET', uri='/tree?token=c0949919ea655c22346df6765b47e8ec323eff55febe3128', version='HTTP/1.1', remote_ip='::1')
    Traceback (most recent call last):
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\web.py", line 1697, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\web.py", line 3174, in wrapper
        return method(self, *args, **kwargs)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\notebook\tree\handlers.py", line 58, in get
        model = cm.get(path, content=False)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 202, in get
        return _run_sync_in_new_thread(get_async)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 909, in _run_sync_in_new_thread
        raise exception
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 900, in _func
        result = IOLoop.current().run_sync(func)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\ioloop.py", line 532, in run_sync
        return future_cell[0].result()
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 200, in get_async
        return (yield _get(self._context(), path, content, type, format))
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 729, in run
        value = future.result()
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 348, in _get
        return (yield GETTERS[(type_to_get, format_to_get)](context, path, content))
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 729, in run
        value = future.result()
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 736, in run
        yielded = self.gen.throw(*exc_info)  # type: ignore
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 359, in _get_file_base64
        return (yield _get_any(context, path, content, 'file', 'application/octet-stream', 'base64', lambda file_bytes: base64.b64encode(file_bytes).decode('utf-8')))
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 729, in run
        value = future.result()
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\gen.py", line 742, in run
        yielded = self.gen.send(value)
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\jupyters3\jupyters3.py", line 376, in _get_any
        last_modified_str = response.headers['Last-Modified']
      File "C:\Users\user\AppData\Local\Continuum\miniconda3\envs\s3\lib\site-packages\tornado\httputil.py", line 239, in __getitem__
        return self._dict[_normalized_headers[name]]
    KeyError: 'Last-Modified'
[E 22:20:23.330 NotebookApp] {
      "Host": "localhost",
      "Connection": "keep-alive",
      "Upgrade-Insecure-Requests": "1",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3",
      "Accept-Encoding": "gzip, deflate, br",
      "Accept-Language": "en-US,en;q=0.9",
      "Cookie": "session=eyJfZnJlc2giOmZhbHNlLCJjc3JmX3Rva2VuIjoiYzljMmU3Mjk2YWY4MmRlMzc2ZDY0ZjczOTQ5ODk1ZTQyNGYxMjllNSJ9.XCzeDg.0jHw5V3sK2KkkQYwyYMFBtabPRo; username-localhost-8888=\"2|1:0|10:1554280375|23:username-localhost-8888|44:MWQ5YzQxNGJhODM5NGQzZmFkYjU5NzllZDg3MTFlNDA=|15699a59f5fb945ded03df926a5906bb0b33ff33ea500d6d5952be110421fc46\"; _xsrf=2|259e6224|a101570435689b3f8e5273c005c43776|1554899840; username-localhost-8889=\"2|1:0|10:1556053060|23:username-localhost-8889|44:ZGM3ZTBhNGQ5NzVkNGZlNGJiM2YzNDY2OWMyMjg4OGM=|eed9b5108797ae55821f891cf7604512e757f50e366dbf6da91a2db7c1477d8f\"; username-localhost=\"2|1:0|10:1556054344|18:username-localhost|44:NjJlZWUxZTBhOWZhNDNlMGIwOTRjMzBmN2IyMTVjMWM=|10479facc69eaa7e6d0e17e0ec3c64e0f98ec2ce82e62466f1c7880cd11562a9\""
    }
[E 22:20:23.334 NotebookApp] 500 GET /tree?token=c0949919ea655c22346df6765b47e8ec323eff55febe3128 (::1) 339.20ms referer=None
```

**Comment by michalc** [comment]
Thanks for the bug report! Will look into it

**Comment by michalc** [comment]
Fixed by https://github.com/uktrade/jupyters3/commit/10494eb40971a3f03f02a6dc1a62d85dee5fc80d and in 0.0.35, https://pypi.org/project/jupyters3/0.0.35/ or https://github.com/uktrade/jupyters3/tree/v0.0.35

The cause:

- So far, we've only used and tested against Jupyter Lab.
- It looks like you're using Jupyter Notebook.
- There seems to be a subtle difference in how the root directory is handled between the two. Lab is `'/'`, but Notebook uses `''` (the empty string).
- The code path followed was then unexpected, resulting in what appeared to be a `HEAD` request against the bucket itself (as opposed to an object), and it looks like the bucket itself does not have a `Last-Modified` header.

---

Closing. If this hasn't fixed your issue, feel free to reply. Thanks!
